nm: Add support of routing rule

cathay4t opened the pull request:
NM does not support run time routing rule query yet.

cathay4t commented:
Wait PR #603

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -23,15 +23,22 @@
+
+# NM require route rule priority been set explicitly, use 3000 when
+# desire state instruct to use USE_DEFAULT_PRIORITY
+ROUTE_RULE_DEFAULT_PRIORIRY = 30000
```
Comment says 3000 and you used here 30,000

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,94 @@ def get_static_gateway_iface(family, iface_routes):
+    cleanup_keys = []
+    for key, val in info.items():
+        if not val:
+            cleanup_keys.append(key)
```
Fits a generator comprehension.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,94 @@ def get_static_gateway_iface(family, iface_routes):
+
+
+def _tuple_to_ip_addr_full(ip_addr, ip_addr_len):
+    return '{}/{}'.format(ip_addr, ip_addr_len)
```
magic: `f'{ip_addr}/{ip_addr_len}'`

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,94 @@ def get_static_gateway_iface(family, iface_routes):
+            ip_addr_len = int(ip_addr_len)
+        except ValueError:
+            raise NmstateValueError(
+                'Invalid prefix length in rule {}'.format(ip_addr_full)
```
The py3 `f''` format will be nicer.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,94 @@ def get_static_gateway_iface(family, iface_routes):
+    return '{}/{}'.format(ip_addr, ip_addr_len)
+
+
+def _ip_addr_full_to_tuple(ip_addr_full, family):
```
Can ipaddress stdlib help you here?

Or maybe it can be generalized somehow and place in iplib module? It shouts "unit test me".

cathay4t commented:
New patch uploaded.

Changes:
 * Added integration test case.
 * Moved ip address related function to `iplib` and benefit from `ipaddress` module:
    * `iplib.to_ip_address_full()`
    * `iplib.ip_address_full_to_tuple()`

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,73 @@ def get_static_gateway_iface(family, iface_routes):
+        RouteRule.PRIORITY: nm_rule.get_priority(),
+        RouteRule.ROUTE_TABLE: nm_rule.get_table(),
+    }
+    cleanup_keys = [key for key, val in info.items() if val is None]
```
If you would have replace the `[]` with `()` it will make it a generator.
But this is a nit, do not bother about it.

cathay4t replied:
When using (), you will get error `RuntimeError: dictionary changed size during iteration`

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -210,3 +217,73 @@ def get_static_gateway_iface(family, iface_routes):
+    ip_to = rule.get(RouteRule.IP_TO)
+    if not ip_from and not ip_to:
+        raise NmstateValueError(
+            'Neither RouteRule.IP_FROM or RouteRule.IP_TO is defined'
```
nit: Shouldn't it be `f'Neither {RouteRule.IP_FROM} or {RouteRule.IP_TO} is defined'` ?

EdDev commented on tests/integration/nm/route_rule_test.py:
```diff
@@ -0,0 +1,158 @@
+        }
+    )
+    libnmstate.apply(state)
+    yield state
```
You do not have a teardown, so a `return` (or nothing) is appropriate.

EdDev commented on tests/integration/nm/route_rule_test.py:
```diff
@@ -0,0 +1,158 @@
+
+
+@pytest.fixture(scope='function', autouse=True)
+def eth1_up_static(eth1_up):
```
`eth1_up_with_static_ip` ?

EdDev commented on tests/integration/nm/route_rule_test.py:
```diff
@@ -0,0 +1,158 @@
+        '2001:db8:b::/64', '2001:db8:1::a/128', 51, 104
+    )
+
+    nm.nmclient.client(refresh=True)
```
Seems to fit a fixture.

EdDev replied:
Can this be now placed in `_assert_route_rules`?

EdDev commented on tests/integration/nm/route_rule_test.py:
```diff
@@ -0,0 +1,158 @@
+    setting_ipv4 = _get_setting_ipv4('eth1')
+    setting_ipv6 = _get_setting_ipv6('eth1')
+
+    nm.route.add_route_rules(
```
Don't you need a mainloop for doing this? We have helpers for that.

cathay4t commented:
@EdDev 

New patch uploaded.

Changes:
 * The integration test will actually apply route rule into OS and then use `ip rule` command to verify that.

EdDev commented:
@cathay4t , if I understand correctly, NM is not able to provide running kernel state, just the profile config. Is this the case?

cathay4t commented:
> @cathay4t , if I understand correctly, NM is not able to provide running kernel state, just the profile config. Is this the case?

Yes.

EdDev commented on tests/integration/nm/route_rule_test.py:
```diff
@@ -0,0 +1,210 @@
+    _check_ip_rules_exist_in_os([rule])
+
+
+def _get_activate_connection():
```
Seems not to be used

EdDev commented on tests/integration/testlib/iprule.py:
```diff
@@ -0,0 +1,69 @@
+        if found:
+            break
+    if not found:
+        logging.debug(f'Failed to found expected ip rule: {expected_rule}')
```
found => find
